// @splitsoftware/splitio-commons/cjs/storages/inMemory/CountsCacheInMemory.js
"use strict";
Object.defineProperty(exports, "__esModule", { value: true });
exports.CountsCacheInMemory = void 0;
var CountsCacheInMemory = /** @class */ (function () {
    function CountsCacheInMemory() {
        this.counters = {};
    }
    /**
     * Add counts.
     */
    CountsCacheInMemory.prototype.track = function (metricName) {
        if (this.counters[metricName] === undefined)
            this.counters[metricName] = 1;
        else
            this.counters[metricName]++;
        return true;
    };
    /**
     * Clear the collector
     */
    CountsCacheInMemory.prototype.clear = function () {
        this.counters = {};
    };
    /**
     * Get the collected data, used as payload for posting.
     */
    CountsCacheInMemory.prototype.state = function () {
        return this.counters;
    };
    /**
     * Check if the cache is empty.
     */
    CountsCacheInMemory.prototype.isEmpty = function () {
        return Object.keys(this.counters).length === 0;
    };
    return CountsCacheInMemory;
}());
exports.CountsCacheInMemory = CountsCacheInMemory;
