// @splitsoftware/splitio/lib/settings/node.js
"use strict";
Object.defineProperty(exports, "__esModule", { value: true });
exports.settingsFactory = void 0;
var settingsValidation_1 = require("@splitsoftware/splitio-commons/cjs/utils/settingsValidation");
var builtinLogger_1 = require("@splitsoftware/splitio-commons/cjs/utils/settingsValidation/logger/builtinLogger");
var LocalhostFromFile_1 = require("../sync/offline/LocalhostFromFile");
var node_1 = require("./defaults/node");
var node_2 = require("./storage/node");
var node_3 = require("./runtime/node");
var params = {
    defaults: node_1.defaults,
    runtime: node_3.validateRuntime,
    storage: node_2.validateStorage,
    logger: builtinLogger_1.validateLogger,
    localhost: function () { return (0, LocalhostFromFile_1.LocalhostFromFile)(); },
    consent: function () { return undefined; }, // resets settings.userConsent to the default
    // In Node.js the SDK ignores `config.integrations`, so a validator for integrations is not required
};
function settingsFactory(config) {
    return (0, settingsValidation_1.settingsValidation)(config, params);
}
exports.settingsFactory = settingsFactory;
