// @splitsoftware/splitio-commons/esm/storages/inMemory/LatenciesCacheInMemory.js
import { findLatencyIndex } from '../findLatencyIndex';
var LatenciesCacheInMemory = /** @class */ (function () {
    function LatenciesCacheInMemory() {
        this.counters = {};
    }
    /**
     * Add latencies.
     */
    LatenciesCacheInMemory.prototype.track = function (metricName, latency) {
        // Initialize if needed
        if (this.counters[metricName] === undefined) {
            this.counters[metricName] = [
                0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0
            ];
        }
        // +1 based on the latency number
        this.counters[metricName][findLatencyIndex(latency)]++;
        return true;
    };
    /**
     * Clear the collector
     */
    LatenciesCacheInMemory.prototype.clear = function () {
        this.counters = {};
    };
    /**
     * Get the collected data, used as payload for posting.
     */
    LatenciesCacheInMemory.prototype.state = function () {
        return this.counters;
    };
    /**
     * Check if the cache is empty.
     */
    LatenciesCacheInMemory.prototype.isEmpty = function () {
        return Object.keys(this.counters).length === 0;
    };
    return LatenciesCacheInMemory;
}());
export { LatenciesCacheInMemory };
